app/jobs/set_collection_job.rb

```diff
@@ -14,43 +14,26 @@ def perform(bulk_action_id, params)
 
     @new_collection_ids = Array(params[:set_collection]['new_collection_id'].presence)
 
-    with_bulk_action_log do |log|
-      update_druid_count
+    with_items(params[:pids], name: 'Set collection') do |cocina_object, success, _failure|
+      check_can_set_collection!(cocina_object)
 
-      pids.each do |current_druid|
-        log.puts("#{Time.current} SetCollectionJob: Starting update for #{current_druid} (bulk_action.id=#{bulk_action_id})")
-        set_collection_and_index_safely(current_druid, log)
-        log.puts("#{Time.current} SetCollectionJob: Finished update for #{current_druid} (bulk_action.id=#{bulk_action_id})")
+      state_service = StateService.new(cocina_object.externalIdentifier, version: cocina_object.version)
+      unless state_service.allows_modification?
+        new_version = open_new_version(cocina_object.externalIdentifier, cocina_object.version, version_message(new_collection_ids))
+        cocina_object = cocina_object.new(version: new_version.to_i)
       end
-    end
-  end
 
-  private
+      change_set = ItemChangeSet.new(cocina_object)
+      change_set.validate(collection_ids: new_collection_ids)
+      change_set.save
 
-  def set_collection_and_index_safely(current_druid, log)
-    cocina_object = Dor::Services::Client.object(current_druid).find
-    state_service = StateService.new(current_druid, version: cocina_object.version)
-    check_can_set_collection!(cocina_object, state_service)
-    unless state_service.allows_modification?
-      new_version = open_new_version(cocina_object.externalIdentifier, cocina_object.version, version_message(new_collection_ids))
-      cocina_object = cocina_object.new(version: new_version.to_i)
+      success.call('Update successful')
     end
-
-    change_set = ItemChangeSet.new(cocina_object)
-    change_set.validate(collection_ids: new_collection_ids)
-    change_set.save
-    Argo::Indexer.reindex_pid_remotely(cocina_object.externalIdentifier)
-
-    log.puts("#{Time.current} SetCollectionJob: Successfully updated #{cocina_object.externalIdentifier} (bulk_action.id=#{bulk_action.id})")
-    bulk_action.increment(:druid_count_success).save
-  rescue StandardError => e
-    log.puts("#{Time.current} SetCollectionJob: Unexpected error for #{current_druid} (bulk_action.id=#{bulk_action.id}): #{e} #{e.backtrace}")
-    bulk_action.increment(:druid_count_fail).save
   end
 
-  def check_can_set_collection!(cocina, state_service)
-    raise "#{cocina.externalIdentifier} is not open for modification" unless state_service.allows_modification?
+  private
 
+  def check_can_set_collection!(cocina)
     new_collection_ids.each do |new_collection_id|
```

Comment:
It would be good to be consistent about whether this should be raised and not caught or handled by invoking `failure.call`.

Reply:
Merged.